src/lib/ledger_catchup/super_catchup.ml

```diff
@@ -392,16 +392,20 @@ let with_lengths hs ~target_length =
 
 (* returns a list of state-hashes with the older ones at the front *)
 let download_state_hashes t ~logger ~trust_system ~network ~frontier
-    ~target_hash ~target_length ~downloader ~blockchain_length_of_target_hash =
+    ~target_hash ~target_length ~downloader ~blockchain_length_of_target_hash
+    ~preferred_peers =
   [%log debug]
     ~metadata:[ ("target_hash", State_hash.to_yojson target_hash) ]
     "Doing a catchup job with target $target_hash" ;
-  let%bind peers =
-    (* TODO: Find some preferred peers, e.g., whoever told us about this target_hash *)
-    Mina_networking.peers network >>| List.permute
+
+  let%bind all_peers = Mina_networking.peers network >>| Peer.Set.of_list in
```

Comment:
What happens if the preferred peers are no longer in the set of connected peers? We should probably prune the preferred peers set each iteration so that we don't keep on trying to contact peers we aren't connected to anymore.

Reply:
I made a fix to do an intersection between the preferred_peers and all_peers to prune the disconnected peers. I don't quite get what you mean by "prune ... each iteration". The loop between just goes over the peers I provide to it. And this loop would only be ran once I think.

Reply:
Sorry, I read the diff for the PR incorrectly and was confused myself. Ignore the detail about "prune each iteration"